cmd/govim/highlight.go

```diff
@@ -86,18 +105,127 @@ func (v *vimstate) redefineHighlights(diags []types.Diagnostic, force bool) erro
 		v.BatchChannelCall("prop_add",
 			d.Range.Start.Line(),
 			d.Range.Start.Col(),
-			struct {
-				Type    string `json:"type"`
-				EndLine int    `json:"end_lnum"`
-				EndCol  int    `json:"end_col"` // column just after the text
-				BufNr   int    `json:"bufnr"`
-			}{string(hi), d.Range.End.Line(), d.Range.End.Col(), d.Buf})
+			propAddDict{string(hi), types.DiagnosticTextPropID, d.Range.End.Line(), d.Range.End.Col(), d.Buf},
+		)
+	}
+
+	v.BatchEnd()
+	return nil
+}
+
+func (v *vimstate) updateReferenceHighlight(refresh bool) error {
+	if v.config.HighlightReferences == nil || !*v.config.HighlightReferences {
+		return nil
+	}
+	b, pos, err := v.cursorPos()
+	if err != nil {
+		return fmt.Errorf("failed to get current position: %v", err)
+	}
+
+	// refresh indicates if govim should call DocumentHighlight to refresh
+	// ranges from gopls since we want to refresh when the user goes idle,
+	// and remove highlights as soon as the user is busy. To prevent
+	// flickering we keep track of the current highlight ranges and avoid
+	// removing text properties if the cursor is still within all ranges.
+	if !refresh {
+		for i := range v.currentReferences {
+			if !pos.IsWithin(*v.currentReferences[i]) {
+				v.removeTextProps(types.ReferencesTextPropID)
+				return nil
+			}
+		}
+		return nil
 	}
 
+	ctx, cancel := context.WithCancel(context.Background())
+	// Cancel any ongoing requests to make sure that we only process the
+	// latest response.
+	v.cancelDocHighlightLock.Lock()
+	if v.cancelDocHighlight != nil {
+		v.cancelDocHighlight()
+		v.cancelDocHighlight = nil
+	}
+	v.cancelDocHighlight = cancel
+	v.cancelDocHighlightLock.Unlock()
+
+	v.tomb.Go(func() error {
+		v.redefineReferenceHighlight(ctx, b, pos)
+
+		v.cancelDocHighlightLock.Lock()
+		if v.cancelDocHighlight != nil {
+			v.cancelDocHighlight()
+			v.cancelDocHighlight = nil
+		}
+		v.cancelDocHighlightLock.Unlock()
+		return nil
+	})
+
+	return nil
+}
+
+func (v *vimstate) redefineReferenceHighlight(ctx context.Context, b *types.Buffer, cursorPos types.Point) {
```

Comment:
This method should be defined on `func (g *govimplugin)`

Reply:
Done